weaviate/collection/classes/internal.py

```diff
@@ -44,10 +89,79 @@ def _metadata_from_dict(metadata: Dict[str, Any]) -> _MetadataReturn:
     )
 
 
-def _extract_props_from_list_of_objects(type_: Any) -> Optional[Any]:
-    """Extract inner type from List[_Object[Properties]]"""
-    if getattr(type_, "__origin__", None) == List:
+def _extract_property_type_from_reference(type_: Reference[P]) -> Optional[P]:
+    """Extract inner type from Reference[Properties]"""
+    if getattr(type_, "__origin__", None) == Reference:
+        return type_.__args__[0]
+    return None
+
+
+def _extract_property_type_from_annotated_reference(
+    type_: Union[
+        Annotated[Reference[P], MetadataQuery], Annotated[Reference[P], MetadataQuery, str]
+    ]
+) -> Optional[P]:
+    """Extract inner type from Annotated[Reference[Properties]]"""
+    if get_origin(type_) is Annotated:
         inner_type = type_.__args__[0]
-        if getattr(inner_type, "__origin__", None) == _Object:
+        if get_origin(inner_type) is Reference:
             return inner_type.__args__[0]
     return None
+
+
+def __create_link_to_from_annotated_reference(
+    link_on: str,
+    value: Union[
+        Annotated[Reference[Properties], MetadataQuery],
+        Annotated[Reference[Properties], MetadataQuery, str],
+    ],
+) -> Union[LinkTo, LinkToMultiTarget]:
+    """Create LinkTo or LinkToMultiTarget from Annotated[Reference[Properties]]"""
+    assert get_origin(value) is Annotated
+    inner_type = value.__args__[0]
+    assert get_origin(inner_type) is Reference
+    metadata = value.__metadata__[0]
```

Comment:
is it guranteed that this always has at least one entry?

Reply:
No, it's a completely free parameter definition on the part of the user. We need to add guardrails around it so that any errors of usage have clear run-time errors thrown. Do you feel these errors are important to get in for the user testing?